src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/repository/ApiKeyRepository.java

```diff
@@ -108,6 +108,44 @@ private String createApiKey() {
         return entry;
     }
 
+    /**
+     * update api key with one role by key id
+     * @param id
+     * @param roleName
+     * @throws InvalidRoleNameException if roleNames was empty or contained invalid roles
+     * @throws ApiKeyNotFoundException if api key was not found
+     */
+    @CacheEvict(allEntries = true, cacheNames = { LIST_ALL_KEYS_CACHE, GET_KEY_CACHE })
+    public PersistentApiKeyType update(long id, String roleName)
+            throws InvalidRoleNameException, ApiKeyNotFoundException {
+        return update(id, Collections.singletonList(roleName));
+    }
+
+    /**
+     * update api key with collection of roles by key id
+     * @param id
+     * @param roleNames
+     * @return
+     * @throws InvalidRoleNameException if roleNames was empty or contained invalid roles
+     * @throws ApiKeyNotFoundException if api key was not found
+     */
+    @CacheEvict(allEntries = true, cacheNames = { LIST_ALL_KEYS_CACHE, GET_KEY_CACHE })
+    public PersistentApiKeyType update(long id, Collection<String> roleNames)
```

Comment:
Feels like this ApiKeyRepository is a bit different from the conventions used in other repositories. Other repositories mostly 
- have a saveOrUpdate method which is used for both creating and updating 
- do not deletegate simple update / delete operations to DAO, but instead just use direct calls with persistenceUtils, e.g. https://github.com/nordic-institute/X-Road/blob/735e853ba2b24871b82606d8df77478ba2283ecb/src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/repository/LocalGroupRepository.java#L104-L106

But this is earlier legacy than this ticket so nothing has to be done here. I think I will add a few lines to confluence to offer some conventions for future repository implementations. 

I think I will pick an API key related ticket next, so I might refactor this part there. Or, if you get the inspiration, you can do it also.

Reply:
Confluence updated: https://confluence.niis.org/display/XRDDEV/API+backend+conventions#APIbackendconventions-Repositories

Reply:
I got inspired and did some refactoring. As a result the implementation now follows the same structure with other endpoints. Repository and service layers are separated, and business logic is implemented in the service layer.

Reply:
That's great! Looks good. Very nice to have this consistent with the rest of the app now.